// messaging-api-slack/dist/SlackWebhookClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const axios_1 = __importDefault(require("axios"));
const ts_invariant_1 = __importDefault(require("ts-invariant"));
const warning_1 = __importDefault(require("warning"));
const messaging_api_common_1 = require("messaging-api-common");
class SlackWebhookClient {
    constructor(config) {
        ts_invariant_1.default(typeof config !== 'string', `SlackWebhookClient: do not allow constructing client with ${config} string. Use object instead.`);
        this.onRequest = config.onRequest;
        // incoming webhooks
        // https://api.slack.com/incoming-webhooks
        this.axios = axios_1.default.create({
            baseURL: config.url,
            headers: { 'Content-Type': 'application/json' },
        });
        this.axios.interceptors.request.use(messaging_api_common_1.createRequestInterceptor({ onRequest: this.onRequest }));
    }
    /**
     * @deprecated Use `new SlackWebhookClient(...)` instead.
     */
    static connect(config) {
        warning_1.default(false, '`SlackWebhookClient.connect(...)` is deprecated. Use `new SlackWebhookClient(...)` instead.');
        return new SlackWebhookClient(config);
    }
    sendRawBody(body) {
        return this.axios.post('', messaging_api_common_1.snakecaseKeysDeep(body)).then((res) => res.data);
    }
    sendText(text) {
        return this.sendRawBody({ text });
    }
    /**
     * Attachments
     *
     * https://api.slack.com/docs/message-attachments
     */
    sendAttachments(attachments) {
        return this.sendRawBody({ attachments });
    }
    sendAttachment(attachment) {
        return this.sendAttachments([attachment]);
    }
}
exports.default = SlackWebhookClient;
